### ISSUE TITLE
Deployment logs for a rolled back deployment

### ISSUE THREAD

**Issue by kargakis** [openshift/origin#9294]
If a deployment config is automatically rolled back by the deployment config controller then `oc logs` will continue returning the logs from the latest failed deployment. It should probably return the logs from the rolled back deployment.

```
[vagrant@localhost sample-app]$ oc status
In project test on server https://10.0.2.15:8443

svc/database - 172.30.67.62:5434 -> 3306
  dc/database deploys docker.io/openshift/mysql-55-centos7:latest 
    deployment #2 failed 2 minutes ago
    deployment #1 deployed 2 minutes ago - 1 pod

https://www.example.com (svc/frontend)
  dc/frontend deploys istag/origin-ruby-sample:latest <-
    bc/ruby-sample-build builds https://github.com/openshift/ruby-hello-world.git with test/ruby-22-centos7:latest 
      build #1 failed 15 minutes ago - e79d887: Merge pull request #53 from danmcp/master (Ben Parees <bparees@users.noreply.github.com>)
    deployment #1 waiting on image or update

Errors:
  * build/ruby-sample-build-1 has failed.

1 error and 3 warnings identified, use 'oc status -v' to see details.
[vagrant@localhost sample-app]$ oc logs -f dc/database
--> pre: Running hook pod ...
--> pre: Success
--> Scaling database-1 down to zero
--> mid: Running hook pod ...
--> mid: Success
--> Scaling database-2 to 1 before performing acceptance check
--> Waiting up to 10m0s for pods in deployment database-2 to become ready
--> post: Running hook pod ...
error: post hook failed: post hook failed, aborting deployment: 
```

**Comment by mfojtik** [comment]
@kargakis i'm not 100% positive we want to display the logs from the previous deployment.

1) The `oc logs dc/mysql` will give you logs from the `LatestVersion` which is the one that failed. I want to be able to debug the failure.
2) In case we will display the 'new-but-previous' logs, I will have to check the pod logs to get the 'failed' one.

To make this work, we will have to record what is the version that is currently running. That means, the rollback will have to set that as Status field (`RunningVersion` or something similar). Then the `logs` will use that version to display logs instead of `LatestVersion`.

Long story short, I would prefer to keep the current behavior. @smarterclayton wdyt?

**Comment by mfojtik** [comment]
@ironcladlou FYI

**Comment by smarterclayton** [comment]
The user does want to debug, but they didn't ask for the latest failure logs, they just said "get me the logs for the DC".  In a rollback, there are pods running.  Displaying the latest deployment pod, regardless of state, means users can't easily see the running pods.

**Comment by mfojtik** [comment]
@smarterclayton the other way around, users will not able to easily see the failure that caused rollback. We were discussing couple options with @kargakis, one of them was introducing the `--active` flag for logs that will cause to display logs for "active" running deployment.

Other option was bump the latest deployment version when rollback happen, so users will be able to see the logs from a running DC and also use `--previous` to see the logs from failed DC. However, this will not just rollback the DC, but cause a new deployment to be rolled out.

Another option would be to show the logs from failed deployment on top, then add marker and continue with streaming logs from the current DC.

**Comment by smarterclayton** [comment]
If I pushed a failure, then rolled back, how long before those failed logs
become irrelevant to me?  After all, I may not care once I've looked once.
Or I may not look after that.  Assuming the user cares about a failed
deployment logs once a rollback has happened doesn't seem like a very
strong argument right now.

**Comment by kargakis** [comment]
@smarterclayton note that the rollback is automatic. I have an issue that we should warn for automatically rolledback configs: https://github.com/openshift/origin/issues/8416

**Comment by smarterclayton** [comment]
We aren't shy about displaying rollback / failure information in the
console and oc status.  To an end user, the failed deployment, while
important, is not as important as their running application.  If there is
something going wrong "right now", the logs for the currently running pods
are more important than the logs for the failed deployment.

**Comment by kargakis** [comment]
I think I agree with the logs for the running application that's why I opened the issue but there are places like `oc status` where we should warn users that their configs are broken and they are one button away from breaking their apps again.

**Comment by kargakis** [comment]
Well, they won't break their apps because minimum availability but anyway I think it's reasonable to warn.

**Comment by kargakis** [comment]
How about showing both and in between let them know what they look at?
```
$ oc logs dc/foo
---> Latest deployment started, hooray
---> bla bla
---> Something broke
---> Failed
== Below are the logs of the previous version of your app
Welcome to foo
logs
logs
```

**Comment by mfojtik** [comment]
@smarterclayton @kargakis will it make sense to implement similar logic as we have with builds, iow. `oc logs dc/foo --previous` which will then display the logs from the previous deployment.

In the log for the current deployment, we can indicate on the first line, that this deployment was triggered as "rollback", something like:

```
$ oc logs dc/foo
-> Latest deployment (#4) for dc/foo
-> WARNING: This deployment was triggered by a rollback of deployment #3, to see the error logs, run: $ oc logs dc/foo --previous
Foo bar bla  # the latest deployment logs...
foo
```

**Comment by kargakis** [comment]
`oc logs dc/foo -p` already shows the previous deployment logs

**Comment by smarterclayton** [comment]
I'm still extremely hesitant to mix those logs right now. It's really
a different experience, and unless users really don't know how to show
previous logs (which we can fix other ways) there isn't a huge reason
to do it now.
